src/librustc_lint/unused.rs

```diff
@@ -354,18 +367,28 @@ impl EarlyLintPass for UnusedParens {
                 }
                 let msg = format!("{} argument", call_kind);
                 for arg in args_to_check {
-                    self.check_unused_parens_core(cx, arg, &msg, false);
+                    self.check_unused_parens_expr(cx, arg, &msg, false);
                 }
                 return;
             }
         };
-        self.check_unused_parens_core(cx, &value, msg, struct_lit_needs_parens);
+        self.check_unused_parens_expr(cx, &value, msg, struct_lit_needs_parens);
+    }
+
+    fn check_pat(&mut self, cx: &EarlyContext, p: &ast::Pat) {
+        use ast::PatKind::{Paren, Range};
+        if let Paren(ref pat) = p.node {
```

Comment:
> I'm not sure btw if the existing unused parens code gets this correct for expressions.

The existing code does get the paren usage correct for expressions. Looking at https://github.com/rust-lang/rust/commit/6399d16cfde37b06f6b82cdafa623e36385d7252, it seems to check specifically for `&pat` or `box pat` and set the `allow_range_pat` flag accordingly. The implementation you've suggested takes this a step further and recurses into a `Vec` of subpatterns which I like.

I agree this would lead to more situations where it _could_ warn, so I'll work on making those changes. Thank you!